scaling.tex

```diff
@@ -1 +1,216 @@
+individual application the communication cost may not be optimal and
+worse, not deterministic.
+
+To try and remove the effect of this variability, a Cray environment
+variable \verb+GNI_BLAS+ was set to $2$ and the $96$ node jobs re-run.
+Whist there was a difference between the run-times, the less adaptive
+version was not consistently faster, nor did it change the balance
+significantly. It should be noted that the communication cost
+node imbalance for the hybrid mode is significantly higher than for MPI
+only. The imbalance (defined as the variation across MPI ranks) is
+around $15\%$ of the halo exchange cost, which rises to $30\%$ of a
+much larger number for the hybrid version. For the global sum, again
+there is a much larger variation for the hybrid version even if the total time spent in
+global sums is reduced compared to MPI only. Changing the Cray
+environment variable settings does not materially change this view. 
+
+Also shown in Table~\ref{tab:scale-data} is the CrayPAT profile data
+for a $96$ node job run with $4$ MPI ranks per node and $9$ OpenMP
+threads per MPI rank. This has a square domain decomposition like the
+MPI only versions. This runs considerably slower than both the $6$
+thread and MPI only version. The OpenMP synchronisation cost barely
+shows in the profile at $1\%$ of run-time. The user code portion runs
+in roughly the same time as the $6$ thread version, which is faster
+than MPI only. However, the communication costs are much higher. Both
+Halo Exchange and Global Sum take much longer and have a high degree
+of imbalance, at around $30\%$ each. The number of Halo Exchanges is
+around 1.3 million, the same as MPI and $6$ thread versions.  It is
+reasonable to believe that at least some of this increase in cost is
+due to network variablility. This could be examined by running in
+large exclusive reservations or with a statistical analysis of a large
+number of runs. Both these are expensive in terms of resource use and
+probably unjustifiable given pressure on resources for science rather
+than performance evaluation.
+
+It is also not obvious why the hybrid version should now be slower
+than MPI when the opposite was true in the FD release. In this data
+the time-step is longer at $180$s and moreover, the FD data was not
+profiled using CrayPAT, but the time for $50$ time-steps calculated
+from the logs. Shown in table~\ref{fig:fd_comp} is a comparison of
+trunk at r17483 and the FD release. Clearly the hybrid version is now
+much slower. The trunk versions uses the new solver API, so the solver
+is a different algorithm but it is not doing a significantly different
+number of halo exchanges. Neither the YAXT library performing the halo
+exchanges nor the generated PSy layer code are any different, {\em
+  i.e.} halo exchanges called outside OpenMP regions, which call halo
+exchange on fields, which call YAXT. The FD model was not profiled
```

Comment:
We've already said that the FD release was not run with CrayPAT

Reply:
Oops. Deleted second CrayPAT statement.